### ISSUE TITLE
Show histogram on top of display range slider

### ISSUE THREAD

**Issue by haesleinhuepf** [Napari/napari#5128]
## 🚀 Feature

I to put a small histogram on top of the display range slider to make it more user friendly. 

## Motivation

When choosing min/max visualized grey value, microsocopists are used to seeing a grey-value distribution (histogram). Pretty much all commercial software has it.

## Pitch

ImageJ can be a good example for a simple solution:

![image](https://user-images.githubusercontent.com/12660498/191953114-5d9c01cf-254f-43e4-8fca-519ae5edfeda.png)

## Alternatives

There is a basic implementation in this napari plugin which might be useful for implementing this in napari-core:
https://www.napari-hub.org/plugins/napari-brightness-contrast

**Comment by Czaki** [comment]
how it should work for zar/dask or other distributed data that may not fit in memory? 
Same as ImageJ where it is only for current, visible layer? Other options?

**Comment by haesleinhuepf** [comment]
> how it should work for zar/dask or other distributed data that may not fit in memory?
Same as ImageJ where it is only for current, visible layer? Other options?

I propose implementing it for common cases first and for special cases such as dask later. It's the user-interface that counts, not compatibility with big data.

**Comment by Czaki** [comment]
> It's the user-interface that counts, not compatibility with big data.

But if we implement histogram calculation for whole data it will break ability to load big data to napari. 
Or you suggest to implement this only for data that fit in memory?

**Comment by haesleinhuepf** [comment]
> Or you suggest to implement this only for data that fit in memory?

Yes. Build the user interface for this first and solve this problem for 99% of the users. Afterwards, one can think about how to do this for big data. Step by step.

**Comment by brisvag** [comment]
I agree this is nice; you can always jsut grey it out for big data. Not sure where to fit it exactly, but for that we need the design people :P

**Comment by haesleinhuepf** [comment]
Maybe @LCObus can assist us or knows someone 🙃

**Comment by Czaki** [comment]
It also will be nice to have the option to set y to log scale.

**Comment by d-v-b** [comment]
+1 for a histogram, and as a big data user I want to point out that there is actually nothing special about that case; as long as napari always shows the distribution of the data being displayed, then big data and small data are essentially the same. Nobody expects the image histogram (which is rendered to a resolution of a few hundred pixels at most) to represent a census of every single value in the data, even for local data.

**Comment by TimMonko** [comment]
At a meeting yesterday (I think our core team meeting, but maybe community) we discussed prioritizing the addition of a histogram to napari for 0.7. I am adding this particular issue to the milestone, despite this potentially not being the implementation we go with (though it is I think the top contender at the moment) -- this will help us track.

**Comment by psobolewskiPhD** [comment]
I spoke at Imaging the Future with @gselzer who mentioned that he recently implemented a histogram in ndv, at least partially based off of some work from @tlambert03 on napari here: https://github.com/napari/napari/pull/675
(look at that PR number, how has this slipped for so long!)
Maybe worth taking a look at things and/or reaching out about pain points/feedback.

**Comment by gselzer** [comment]
Thanks for the ping @psobolewskiPhD! Getting some histogram functionality in napari would be a huge step forward!

I would be more than happy to discuss things if anyone thinks I can be of help!

**Comment by DarioS** [comment]
Could there also be text boxes on the left and right side of the slider for precise setting? I am thinking about hot pixels.
```r
> table(imageData(all[["Patient45"]])[, , "CD20"])
      0       1       2       3       4       5       6       7       8       9      10      11  10157
1142819  215001   57893   22191    9949    4958    2456    1227     668     341     201     112      1
```
That one pixel with intensity 10157 makes the slider unusable for choosing a maximum value amongst 8, 9, 10, 11.

**Comment by TimMonko** [comment]
@DarioS, great question! This should be possible now, but I'm a sneaky way. If you double click the displayed numeric values, it should become an editable box for which you can enter any n number. Thank you for the reminder that this is not an obvious feature

**Comment by Czaki** [comment]
The big problem here is cost of histogram calculation. 

We have a special code for estimate contrast range for bigger data, when histogram calculation on smaller one is still expensive. 

We should have this in mind when implement this.

**Comment by psobolewskiPhD** [comment]
One option would be to only implement the histogram for reasonably small images and e.g. numpy arrays only.
Another is to look for acceleration libraries, e.g. https://github.com/marktsuchida/ihist
